### PR TITLE
Optimize query

### PR DESCRIPTION
**Pull request by s-taylor**
This PR adds...
- Refactor controller code to be more DRY (move code into libs shared by API requests and Sockets)
- Automatic query optimisation using table indexes.
- Adds tests for query optimisation
- Add tests for .getIndex(x)
- Refactor the file scaffolders (base/index.js and controller/index.js)

Follow up PRs
- Move scaffold code to render mustache and generate folders/files into separate lib (code is identical)
- Optimise sortBy using indexes (rethinkdb expects different syntax).
- get rid of transformations (limit, skip and sortBy) from sockets, doesn't make sense to me keep these, currently in discussion in Flowdock with Dave.

### CODE REVIEW

**Comment by nwinch** [comment]
@s-taylor Are you adding the "follow up PRs" as issues?

File: controller/controller.mustache [hunk 1]
```diff
@@ -35,18 +41,28 @@ module.exports = {
...
+    const indexBy = controller.getIndex(modelSchema, queryParams);
 
-    if (params.id) {
-      return table.get(params.id).changes({includeInitial: true, includeStates: true}).run();
+    var query = rethinkdb.buildQuery(table, indexBy, queryParams);
+    query = rethinkdb.addTransformations(query, params);
+ 
+    //if query params were passed but none were on an index field
+    if (!_.isEmpty(queryParams) && indexBy == null) {
+      console.log(NOT_INDEXED);
```
Comment (nwinch) [inline controller/controller.mustache]: Remove

File: controller/controller.mustache [hunk 2]
```diff
@@ -35,18 +41,28 @@ module.exports = {
...
+      query = table.getAll(r.args(query.getField('id').coerceTo('array')));
     }
 
-    params = _.assign({order: 'asc'}, controller.normaliseParams(params));
-
-    const filteredTable = table.filter(schema.filter('{{name}}', params));
-    const query = rethinkdb.buildQuery(filteredTable, params);
+    if (params.limit || params.skip) {
+      console.log(SORT_LIMIT);
```
Comment (nwinch) [inline controller/controller.mustache]: Remove

File: controller/lib/controller.mustache [hunk 3]
```diff
@@ -10,3 +12,17 @@ exports.normaliseParams = params => {
...
+exports.getIndex = (schema, params) => {
+  const fields = Object.keys(params);
+  if (params.id && fields['id'] && fields['id'].indexed) return 'id';
+
+  const properties = schema.properties;
+  for (var key in params) {
+    if (params.hasOwnProperty(key)) { 
+      if (properties[key] && properties[key].indexed) return key;
+    }
+  }
```
Comment (nwinch) [inline controller/lib/controller.mustache]: You've added `lodash` at the top so you could take advantage of it with this loop, with something simple like:
```
_.findKey(params, (val, key) => properties[key].indexed)
```

**Comment by nwinch** [comment]
@s-taylor Review done with a couple of pieces of feedback.

File: controller/controller.mustache (hunk 1 above)
Comment (s-taylor) [inline controller/controller.mustache]: As discussed going to leave this as is as was Dave's recommendation to warn the user via console.log. I also think it'd be good to have a better way to do this, but not sure what that would look like. We could emit a warning via sockets or something, but we'd then have to do something with it :/

File: controller/controller.mustache (hunk 2 above)
Comment (s-taylor) [inline controller/controller.mustache]: As above

**Pull request merged**